pkg/controller/windowsmachineconfig/tracker.go

```diff
@@ -79,14 +75,112 @@ func newNodeRecord(ipAddress string, secretName string) nodeRecord {
 	}
 }
 
+// GetWindowsVM returns a WindowsVM interface for the specified VM
+func GetWindowsVM(instanceID, ipAddress string, credentials Credentials) (types.WindowsVM, error) {
+	winVM := &types.Windows{}
+	windowsCredentials := types.NewCredentials(instanceID, ipAddress, credentials.Password, credentials.Username)
+	winVM.Credentials = windowsCredentials
+	// Set up Winrm client
+	if err := winVM.SetupWinRMClient(); err != nil {
+		return nil, errors.Wrap(err, "error instantiating winrm client")
+	}
+	// Set up SSH client
+	if err := winVM.GetSSHClient(); err != nil {
+		return nil, errors.Wrap(err, "error instantiating ssh client")
+	}
+	return winVM, nil
+}
+
+// GetNodeIP gets the instance IP address associated with a node
+func GetNodeIP(nodeList *v1.NodeList, instanceID string) (string, error) {
+	// Ignore the nodes that are not ready.
+	for _, node := range nodeList.Items {
+		for _, condition := range node.Status.Conditions {
+			if condition.Type != v1.NodeReady {
+				continue
+			}
+		}
+		if strings.Contains(node.Spec.ProviderID, instanceID) {
+			for _, address := range node.Status.Addresses {
+				// If external ip exists return it as it is used in AWS, if it doesn't return internal(in case of azure)
+				// TODO: collect all the ips and define a priority order, external then internal etc.
+				if address.Type == v1.NodeExternalIP && len(address.Address) > 0 {
+					return address.Address, nil
+				}
+			}
+		}
+	}
+	return "", errors.Errorf("unable to find Windows Worker node for VM with instance ID %s", instanceID)
+}
+
+// getCredsFromSecret gets the credentials associated with the instance.
+func getCredsFromSecret(k8sclientset *kubernetes.Clientset, instanceID string, operatorNS string) (Credentials, error) {
+	creds := Credentials{}
+	instanceSecret, err := k8sclientset.CoreV1().Secrets(operatorNS).Get(instanceID, metav1.GetOptions{})
+	if err != nil && k8serrors.IsNotFound(err) {
+		return creds, errors.Wrapf(err, "secret asociated with instance %s not found", instanceID)
+	}
+	if err != nil {
+		return creds, errors.Wrap(err, "error while getting secret")
+	}
+	encodedCreds := instanceSecret.Data[instanceID]
```

Comment:
Do you need to check if the key exists in the map? Or will that be safely caught in the unmarshal?

It might be nice to have a specific error message if the key is missing in any case.

Reply:
There 2 assumptions here:

- Secret name should match the entry in it's Data field key.
- As you suggested, it will be caught later in unrmashaling. 

However I don't mind adding a check